### ISSUE TITLE
Wishlist

### ISSUE THREAD

**Issue by PhilterPaper** [sciurius/perl-HarfBuzz-Shaper#3]
I've been playing with HarfBuzz::Shaper in conjunction with PDF::Builder, and it looks promising. As is, it appears to be handling Arabic, Devanagari, and Khmer properly! I was afraid that I would need to upgrade it quite a bit to handle these complex scripts, but it seems to be guessing the script parameters correctly (at least, in these limited cases). Attached is a test page [testHS.pdf](https://github.com/sciurius/perl-HarfBuzz-Shaper/files/4143272/testHS.pdf) from [testHS.pl](https://github.com/sciurius/perl-HarfBuzz-Shaper/files/4143291/testHS.pl.txt) that shows off some stuff. The first line is the raw `text()` output, and the second line is Harfbuzz::Shaper output. For Latin alphabet texts, it looks like it _might_ be doing some kerning, but I could do with it closing up AWAY some more than it does. 

I'm disappointed that it does not appear to be doing ligatures, so I'm wondering if it would be useful to be able to override the "best guess" settings to specify (if needed), the language (en), the script (Latn), and the direction (LTR). I take it this would require a new call in the XS glue, but that should be minor. I might also have to be able to specify features such as 'liga' (do ligatures), 'cpsp' (kerning), 'smcp' (small caps), etc. According to [this article](https://lwn.net/Articles/741722/), I should be able to add a string like ":liga:cpsp:smcp" "after the font name", but I can't find where to do it or any other documentation (it doesn't work adding it to the font filename in `set_font`). There seems to be an example in the HarfBuzz documentation to request 'dlig', so it's possible that another function will have to be added to the XS glue and interface.

A few notes about the sample PDF: I have added some code to PDF::Builder (not yet in library) to accept streams of CIDs (4 digit hex codes) returned by Shaper, rather than Unicode points. If it looks useful, I'll put it in the library. Second, I had to tell PDF::Builder to embed the entire font, and not just the subset. In the Khmer example, the little "n" below other characters was specified in the PDF, but didn't show up when the embedded font was subsetted. That needs to be investigated (I think Johan mentioned a problem that might be the same thing, about having to force characters into the subset). Finally, Shaper returns characters one at a time, with offsets and advances. Right now I'm doing a `translate` and write one character at a time, which really bloats the PDF. If there's only x_advance, I might be able to use `TJ` and group a lot of characters together, but I'm concerned that might mess up kerning and might not work that well with bidirectional (RTL) text and vertical text. I'm open to suggestions on how this might best be done.

So, how does this interact with Text::Layout's planned use of HarfBuzz::Shaper? If I can get all this stuff done through Text::Layout, I will probably choose to just use Text::Layout. However, I'm concerned about future needs on two fronts, and whether Text::Layout + HarfBuzz::Shaper will be up to the job: (1) full support for non-Western and complex scripts, and (2) the ability to do paragraph shaping, including pouring text into irregular columns (with column balancing) and justified lines. If Johan wants to bundle all this into his library, that would be OK by me, but I'm starting to worry about so much of this disappearing into one library that PDF::Builder will simply lose control over what it's doing. 

Perhaps it would be better to have Text::Layout return a data structure for a big chunk of text (that could be concatenated with others), with tentative locations, chunk sizes, and font information (having called HarfBuzz::Shaper to get glyph information once the font selection and font sizing is established). It could then call a page shaper (Knuth, et al.) to arrange for a pleasing layout, and somewhere in there pouring of text into multiple balanced columns (that might not be  boring rectangles), if desired. Somewhere in there, line splitting/hyphenation needs to be done. When all the sizing and positioning is done, the page would be finalized by actually outputting it to the file. That's what I'd like to be able to do with PDF::Builder, but would require that I not hand the whole thing over to Text::Layout, and have to simply accept the finished PDF at the end.

**Comment by sciurius** [comment]
Good to see that you've got it going!

The PDFAPI2 backend for Text::Layout is used for both PDF::API2 and PDF::Builder. At the end of the backend file there are two 'helper' routines for PDF::API2, I'll add these for PDF::Builder as well. These helper routines solve the issues of glyph printing and font subsetting.

HarfBuzz::Shaper currently uses the library call hb_buffer_guess_segment_properties to guess the properties. Additional calls can be added for explicit specification of properties. PRs are welcome.

The page shaper idea is nice, I think Text::Layout (or its algorithms) can play a role in that. Personally (and ultimately) I would like PDF::API2/Builder to provide this functionality and consider Text::Layout as a first (external) step. I do, however, see a separation between the low-level functionality of PDF::API2/Builder and higher level paragraph/page formatting. It may be beneficial to put the higher level functionality in a distinct module instead of integrating with PDF::API2/Builder.

**Comment by PhilterPaper** [comment]
Is there any chance of separating x_advance into x_charwidth and x_adjust? Likewise for y. If done that way, if x_adjust, y_adjust, x_offset, and y_offset are all 0 I could just merge the glyph CID strings into one long one for a TJ operator or even a Tj, with occasional breaks for kerning adjustments, etc. The idea is that right now I have to individually place and output just one glyph at a time, which is terribly inefficient. Do your "glyph printing" routines help with this? BTW, I fixed the issue with subsetting, so now the PDF just carries the subset of glyphs.

HarfBuzz allows you to turn on and off many "features"; that would be a good thing to consider. Specifically, I need ligatures (e.g., f-f-l into ffl) but it doesn't appear to be using them. I asked specifically for language, script, and direction property input (optionally override guess_segment_properties) because I _suspect_ that perhaps it's using DEFAULT rules rather than ENGLISH, and avoiding doing ligatures (which may be language dependent). BTW, the standard C++  header include files defining constants for these calls would presumably not be accessible from Perl. They probably don't change that often, but at the Perl level new ones would have to be defined, or just document which bare integers to use. Perhaps even a utility to read c/include/harfbuzz and generate Perl constants code, if there's any chance of the HarfBuzz library changing out of synch with HarfBuzz::Shaper? Anyway, if it sounds to you like this might get feature support off the ground, I could try coming up with PRs.

**Comment by PhilterPaper** [comment]
My request to pull back on the integration into Text::Layout (perhaps the discussion should go on over there?) stems from the inflexibility introduced by feeding my text into Text::Layout and (without any further control by me) it appears in the PDF. There are things I will need to do in the future for page-level layout that would be much easier if you returned control to me after each step, rather than doing it all in Text::Layout. And yes, at some point routines invoking PDF::Builder as only a low/medium level library might be better than putting it all in PDF::Builder.

Given some text with high (block) level and low (inline) level formatting, a new page shaping library (PDF::Builder::PageLayout?) could handle high level formatting, and pass on remaining text and formatting to Text::Layout. Instead of outputting to the PDF, Text::Layout would return a data structure with font changes, chunk widths and heights (known from HarfBuzz::Shaper), etc. for a long stream of text, such as a full paragraph. This would then be fed into a proper paragraph shaping call (Knuth, et al.) to get relative placement and line-breaking (somewhere, words will have to be split). In conjunction with that, formatted text needs to be poured into columns, which may not be uniform rectangles. The dividing line between parts is still a bit fuzzy -- for example, you can't really shape a paragraph in isolation, because it may cause an orphan in the following paragraph (not only the first line of text, but also any headings). I suspect that it may not be a straight-through or waterfall process, but will have some iterations. More than one page of output may have to be held in memory until it is finalized. Finally, when all text properties and placement are stabilized, this data structure would be fed to an output routine to call the low-level PDF::Builder routines to actually output at least one page of the PDF. The tail end of the buffer might be held in case further output requires adjustments to the text placement.

Anyway, that's a high level overview of what I think this thing will eventually become. Unless you wish to handle it all in Text::Layout, it would probably be better to make all the stages accessible to user code.

**Comment by PhilterPaper** [comment]
Ligatures, in theory, should be independent of the language of the text, and depend only on the font (typeface and attributes) being used. Originally, ligatures were created to solve a problem of type breaking due to collisions between certain letters (especially **f** followed by **f**, **i**, or **l**). They are not part of any language itself, unless you want to count sz, tz, AE, OE, etc., but these are generally treated as their own characters. For mainly stylistic reasons, a number of other ligatures may be supported by a font. With hot metal type (Linotype, Monotype, etc.), or even later, with screens and laser printers, type breakage is no longer a concern, but readers still think it looks classy to use ligatures, so they are still provided.

There _is_ considerable debate in typesetting circles about _when_ you should use a ligature, and when you should use separate letters. All agree that in the source text, and searchable text within a PDF, ligatures should be spelled out as separate letters (for typographic ligatures, not characters such as eszett **sz**). The argument comes over whether you should use, say, the **fi** ligature _every_ time you encounter **f-i**, or in some cases just leave it as **f-i**. One group says that if the font designer has included a ligature, it should be used (consistently), while the other says that a ligature should not span morpheme boundaries, as it confuses the meaning of the word.

A morpheme is kind of hard to describe, but could be thought of as a word atom. In the word "shelfful" (I have a shelfful of books), "shelf" and "ful" are the morphemes (-ful is a suffix). The "no span" group is concerned that tying together **f-f** into a ligature confuses the reader and slows them down, as it's harder to visually split the word into "shelf" and "ful" and understand that the writer is talking about a bookshelf that is full of books. Likewise, "bluff" is a single word and a single morpheme, so the **ff** ligature would be acceptable, but "bluffing" is two morphemes, and while an **ff** ligature would be fine, any available **ffi** ligature should be avoided because it spans morphemes.

These days no one (except artisans using cold metal type) needs worry about breaking type, so there's no harm in avoiding ligatures. The question is whether _using_ them in all cases (where they span morphemes) introduces confusion to the reader and slows them down, or whether visual consistency is more important. BTW, the preferred way of preventing a ligature is to introduce a ZWJ (zero width joiner) character between letters of the possible ligature: "bluff\ZWJing".

**Comment by sciurius** [comment]
> Is there any chance of separating x_advance into x_charwidth and x_adjust?

HarfBuzz::Shaper tries to be as thin as possible. The harfbuzz library call `hb_buffer_get_glyph_positions` yields ax, ay, dx and dy values. The HarfBuzz::Shaper method `shape` delivers the information it gets from harfbuzz. 

> The idea is that right now I have to individually place and output just one glyph at a time, which is terribly inefficient. Do your "glyph printing" routines help with this?

No, it is just a single-glyph `Tj`. However, I do not worry about efficiency in this stage. You know that "Premature optimisation is the root of all evil" (Hoare/Knuth).

> Specifically, I need ligatures (e.g., f-f-l into ffl)

I am not in favor of enforcing ligatures at this level. Ligatures are language and context dependent and useful only in very few cases. It is dead wrong to turn f-f-l into ligature ffl unless you know what the language and context is.

Having said that, library calls to control harfbuzz features will be added in due time, when required.

> I could try coming up with PRs.

Always welcome.

**Comment by sciurius** [comment]
> Given some text with high (block) level and low (inline) level formatting, a new page shaping library (PDF::Builder::PageLayout?) could handle high level formatting, and pass on remaining text and formatting to Text::Layout. Instead of outputting to the PDF, Text::Layout would return a data structure with font changes, chunk widths and heights (known from HarfBuzz::Shaper), etc. for a long stream of text, such as a full paragraph.

This is basically what happens. You feed texts (with markup, font changes, etc) to Text::Layout and it will calculate the glyphs and coordinates. When using the PDFAPI2 backend these are then fed into the lower level PDF::API2/Builder routines. It is straighforward to add another backend that just returns the details to the caller.

> Anyway, that's a high level overview of what I think this thing will eventually become.

Yes, that would be nice.

> Unless you wish to handle it all in Text::Layout, it would probably be better to make all the stages accessible to user code.

See the first paragraph,

**Comment by sciurius** [comment]
> These days no one (except artisans using cold metal type) needs worry about breaking type, so there's no harm in avoiding ligatures.

Good.

> BTW, the preferred way of preventing a ligature is to introduce a ZWJ (zero width joiner) character between letters of the possible ligature: "bluff\ZWJing".

I hope you don't seriously expect users to start writing texts like that.

The added value of good ligatures is next to zero except for very few cases. So let the exception be the exception.

**Comment by PhilterPaper** [comment]
You could always disable (or refuse to enable) 'liga' automatic ligature substitution for your own productions, if you don't like ligatures in your output text. I think that many other users _would_ like the ability to use ligatures, and if it's not a huge amount of work to implement, the ability should be there.

If a user wants to generally use ligatures, but selectively omit them in certain places (in the "don't cross morpheme boundaries" camp), the traditional way to do this (see TeX and maybe earlier) is to mark those cases as "no ligature here". Yes, it is clumsy to have to manually insert `\x{200d}` (or maybe a `\NL` that gets processed to `\x{200d}`) in text strings, so maybe something can be done to automate it. One could keep a language-dependent list of words to be so adjusted in text, and either preprocess the whole source file, or do it on the fly word-by-word in the `text()` routine (user hook to the list of words). Another way, once paragraph shaping is in place, is to borrow paragraph shaping's word-splitting (for line breaking hyphenation) to see where the syllables are. The problem with that is a morpheme is not necessarily a syllable, and vice-versa, so some ligatures may be unnecessarily suppressed. E.g., I think "waffle" is waf-fle, which would forbid an **f-f-l** ligature. Anyway, enabling ligatures, with `\x{200d}` manual suppression, could do for the time being. If you want to _suppress_ ligatures in general, but use one once in a while, some mechanism would have to be provided to do that (you don't want to put the Unicode ligature point in the original text, as it's then unsearchable).

I will have to look at Text::Layout to study the backend structure and see about returning a data structure at perhaps several stages. 

1. Turn font changes (including italics, bold), size changes, color/underline/strikethrough into a data structure with the associated text chunks.
2. Use Knuth et al. algorithms to find break points in words.
3. Call HarfBuzz::Shaper to get the width and height of each chunk (now glyph list).
4. Pour text into columns (not necessarily simple rectangles) and call Knuth-style paragraph shaping to decide how to lay out each paragraph (assign tentative positions relative to the paragraph start, for each chunk).
5. Go back and attempt to adjust earlier paragraphs to avoid orphaned headings and lines (on _this_ paragraph), and any other high-penalty problems in this paragraph.
6. Once you have several pages in the buffer, and the oldest page has stabilized, process its data structure to finally render it with PDF::Builder and actually output it.
7. Repeat for next paragraph.

The result: beautifully formatted pages! I hope. Obviously this is more than Text::Layout itself, and even PDF::Builder, should handle by themselves.

Regarding output of HarfBuzz::Shaper results, simply outputting multiple `<nnnn> Tj` isn't going to work. If ay==dx==dy==0 and ax is equal to the character width, that would be OK (as would merging multiple glyphs into one `<...> Tj` string), but if kerning or other adjustments are included in ax, the output will be wrong. If x_character_width and x_adjust were separate, a TJ array could be used. Of course, ay, dx, or dy not 0 would toss everything out the window and force each glyph to be individually positioned (Tm), which really bloats the PDF. I'm still hoping to find something in the PDF spec that allows a more compact representation.

**Comment by PhilterPaper** [comment]
Over in Text::Layout land, https://github.com/sciurius/perl-Text-Layout/issues/6 asks for font substitution when the current font fails to include a desired character. Assuming that HarfBuzz tells us in some way that a glyph is missing, is this something that might be better done at the HarfBuzz::Shaper level? If another font needs to be brought in, my understanding is that this chunk of text would then need to be split so that Shaper is only handling a single font (and size) at a time. This also brings up the concept of redoing an entire _word_ at a time, so that the font doesn't change in the middle of a word. That could get complicated if the new font is missing any characters in the original font! Anyway, it's another thing to ponder when thinking about overall page shaping.

Using a different list of fonts for a range of characters could still be done in Text::Layout, but this concern is over handling fallbacks when a character isn't found in the primary font. And how do you pass on the list of fonts to Shaper?

**Comment by PhilterPaper** [comment]
Something to add to the wishlist, triggered by the previous post: when you change font size or the font itself in the middle of a line, you could end up with different ascender and descender heights. The result (if either one is larger) is that you need to shift the text baseline (for the **entire** line) so that leading remains correct and the taller/deeper text doesn't collide with adjacent lines. This would mean knowing the overall height of a chunk of processed text (and length, while you're at it), and where the baseline is. A line therefore cannot be output to the PDF (or other) file until it is complete and it is known where the baseline will be. Positioning of the following line has to know how deep the line above it has gone, so as to avoid collisions.

**Comment by PhilterPaper** [comment]
I've been trying a few things, and having problems. 

1. I can't get get_language() and set_language() to work. I assume that you tested these before release, so I must be using them incorrectly. get_language() always returns "english-united", whether or not I had called set_language(). I called "set" just before the shaper() call, and "get" just after. "set" argument such as "fr-FR" was still returned ("get") as "english-united".

2. I added "get_script/set_script" and "get_direction/set_direction" modeled on your "get_language/set_language" routines. See [shaper.txt](https://github.com/sciurius/perl-HarfBuzz-Shaper/files/4534508/shaper.txt) for the code. "get_direction" seems to work OK (returns "ltr" or "rtl" as expected), but "set_direction" seems to be ignored. This one is important so that someone could write Chinese or Japanese in the customary style of vertical lines (set 'ttb').

3. "get_script/set_script" isn't so important (more to round out the expected set), but "get_script" blows up Perl (I haven't tried "set_script" yet). I was wondering how far you got with get/set of direction and script (when you did the language calls), and if the attached code is of any use.

Other than that, I released PDF::Builder 3.018 yesterday, including a lot of support for HarfBuzz::Shaper. I'm hoping that just a few last little items would make it complete. Hopefully it's mostly a matter of beefing up the documentation with exactly how to use a call. Thanks!

**Comment by khaledhosny** [comment]
> Is there any chance of separating x_advance into x_charwidth and x_adjust? Likewise for y. If done that way, if x_adjust, y_adjust, x_offset, and y_offset are all 0 I could just merge the glyph CID strings into one long one for a TJ operator or even a Tj

This can be achived by getting the gllyph advance from the font (e.g. by `hb_font_get_glyph_h_advance()`) and compare it to the `x_advance` from glyph info.

**Comment by khaledhosny** [comment]
> Second, I had to tell PDF::Builder to embed the entire font, and not just the subset. In the Khmer example, the little "n" below other characters was specified in the PDF, but didn't show up when the embedded font was subsetted.

The subset should use the glyphs returned from the shaper not the input characters.

**Comment by PhilterPaper** [comment]
> This can be achived by getting the gllyph advance from the font

That's what I ended up doing... getting the glyph width from the font file, and comparing it to the 'ax' element to see how much kerning +/- was built in. It would have been a lot nicer if HarfBuzz handled this, but I'll take what I can get.

> The subset should use the glyphs returned from the shaper

It does. The problem turned out to be that there were some code paths where some GSUB'd glyphs were not being added to the subset list (PDF::Builder). It's been fixed.

**Comment by sciurius** [comment]
Phil, can you try this to see if language, script and direction work as you'd expect?

[HarfBuzz-Shaper-0.023.tar.gz](https://github.com/sciurius/perl-HarfBuzz-Shaper/files/4750938/HarfBuzz-Shaper-0.023.tar.gz)

**Comment by khaledhosny** [comment]
> The default language (**get** only) is still "english-united",

If no language is set and `hb_buffer_guess_segment_properties()` is called (which is the case here), HarfBuzz will set the language from the current locale so that is where it is getting "english-united".

> If I _set_ the language to "fr-Fr", the _get_ returns "fr-fr".

HarfBuzz normalizes the language tags, and part of that is lowercasing them.

**Comment by PhilterPaper** [comment]
So "current locale" includes "english" and "united states" in my case, and is "C" in Johan's case? @sciurius do you explicitly set your **locale** to "C" on your system? Any suggestions on how to turn this into a more standard representation, or will we just have to accept that without a 'set', the 'get' language value is probably useless?

A _language_ is supposed to be lowercase, but the second part ('FR' or 'DE') is a _country_, and it sounds to me like BPC-47 says it should be uppercase. Is it a mistake for HarfBuzz to fold it to lowercase?

**Comment by khaledhosny** [comment]
On Unix systems, `locale` should tell your current locale. But in general yes, depending on `hb_buffer_guess_segment_properties()` is bad as it does a very simplistic guessing job, proper script, direction and language need to be explicitly set by the caller.

It does not matter if the country code is uppercase or lowercase, HarfBuzz normalizes to lowercase and works with that. Country code does not even matter most of time, since the language will be mapped to [OpenType language tags](https://docs.microsoft.com/en-us/typography/opentype/spec/languagetags) and only a few of them differentiate by country.

**Comment by PhilterPaper** [comment]
@sciurius it appears that you incorporated my get/set script and direction code, and that seems to be mostly working! (I haven't tested set_script yet) The only fly in the ointment seems to be that if I set, say, some Chinese text to "TTB" direction, all cases _after_ that are now TTB, even if naturally they would be LTR or RTL. It appears that something is not being reset internally when I load up the next buffer. Same goes, for, say, Arabic text leaving the state as RTL even when I load in an English phrase (Latn script). Is that an easy fix, or is there a call to reset the buffer when I go to the next chunk? I just want to avoid, if I can, destroying and creating a fresh buffer each time, unless that's the best way to do it.

Once this is ironed out, I will have to get TTB and BTT text directions implemented in textHS().

**Comment by sciurius** [comment]
Yes, I mostly copied your code. Thanks for providing this.

I'm not sure what you mean by 'next buffer' and 'next chunk'. The current implementation is to collect all pieces of text (`set_text`) and apply all when `shaper` is called. Each `shaper` call expects the text to be of same font/size/language/direction/script. This is in accordance with the semantics of the underlying `hb_shape` function of the harfbuzz library.

I consider creating/resetting a harfbuzz buffer a light-weight operation. The real work is in the `hb_shape` call, and caching is in the font data structures.

**Comment by PhilterPaper** [comment]
Looking promising. I called `reset()` just before `set_direction('TTB')` and it seems to reset the direction OK. I think the next step is my updating `textHS()` (PDF::Builder) to properly handle TTB and BTT, and see how it goes then. I've got some stuff going on right now, so this may take several days. I'll let you know how it turns out.

**Comment by sciurius** [comment]
Great. I think it is safe to revert the change of `hb_buffer_clear_contents` to `hb_buffer_reset` since that wasn't related to the problem.

[HarfBuzz-Shaper-0.023.tar.gz](https://github.com/sciurius/perl-HarfBuzz-Shaper/files/4762896/HarfBuzz-Shaper-0.023.tar.gz)

**Comment by sciurius** [comment]
Just wondering...

If you have LTR text `abc` and `def`, and RTL texts `uvw` and `xyz`, what would be the expected result of

````
$hb->set_direction('LTR');
$hb->set_text('abc');
$hb->set_direction('RTL');
$hb->set_text('uvw');
$hb->set_text('xyz');
$hb->set_direction('LTR');
$hb->set_text('def');
````
Would it be `abcwvuzyxdef` or `abczyxwvudef`? Or am I talking nonsense?

**Comment by PhilterPaper** [comment]
I haven't tried multiple `set_text()` in a row. It would be reasonable to _either_ concatenate them first (resulting in uvwxyz -> zyxwvu) or process them first and then concatenate into wvuzyx. You would just have to document which way.

If all these are _before_ the `shaper` call, do you expect it to preserve multiple directions, or isn't it supposed to have all the text in the buffer going the same direction? Are you implying that you are calling `shaper` after each `set_text`?

Same results (OK) as the latest one last night, with your latest install image.

**Comment by sciurius** [comment]
Each `set_text` overrides current content, so there must be shaper() calls in between:
````
$hb->set_direction('LTR');
$hb->set_text('abc');
$hb->shaper();
$hb->set_direction('RTL');
$hb->set_text('uvw');
$hb->shaper();
$hb->set_text('xyz');
$hb->shaper();
$hb->set_direction('LTR');
$hb->set_text('def');
$hb->shaper();
````

**Comment by PhilterPaper** [comment]
OK, I _thought_ there might be `shaper` calls. In that case, my expectation would be that 'uvw' comes out as 'wvu' and then 'xyz' **defaults** to 'xyz'. I.e., the direction gets reset to default each time, unless you explicitly `set_direction`. 

If someone has a good reason for an explicit `set_direction` to _keep_ overriding the default direction until the next `set_direction`, I'm willing to listen to their explanation. But at this point, I think that behavior should reset each time (how Shaper used to be), regardless of whether it was an explicit or default setting. If I embed some LTR text within a RTL sentence (Arabic or Hebrew), I think the natural output should be RTL LTR RTL. I shouldn't have to explicitly call `set_direction` for any chunk. I don't think `set_direction` should have to be called all that often, except for things like TTB and BTT output for East Asian languages. If someone wants to embed some Latin (LTR) script within it, they would have to explicitly call `set_direction` for each chunk.

Perhaps you could consider adding some sort of "global override" versions of set_* calls, to keep going until explicitly canceled or overridden with a new set_* call? That way, someone printing Chinese or Japanese in the traditional (TTB) manner could embed English or other LTR text without worrying about it coming out LTR in a column (because they forgot to explicitly override the natural LTR). Vertical layout is something of an odd duck -- when mixing RTL and LTR, I think automatic setting of the direction would be the most natural.

**Comment by sciurius** [comment]
> If I embed some LTR text within a RTL sentence (Arabic or Hebrew), I think the natural output should be RTL LTR RTL. I shouldn't have to explicitly call set_direction for any chunk.

But what if the sentences contains different typefaces (bold, italic, in sofar that exists in RTL languages)? 
````
set_direction('LTR');
set_font(...regular...);
set_text(...);
shaper();
set_font(...italic...);
set_text(...);
shaper();
set_font(...regular...);
set_text(...);
shaper();
set_direction('RTL');
set_font(...regular...);
set_text(...);
shaper();
set_font(...italic...);
set_text(...);
shaper();
set_font(...regular...);
set_text(...);
shaper();
````
and so on.

The current API is consistent in that _all_ settings persist until overridden or `reset`.

**Comment by khaledhosny** [comment]
I suggest first dropping the call to `hb_buffer_guess_segment_properties()` as it encourages sloppy code, if you really want to let users use it make calling it explicit decision by the user. `hb_buffer_guess_segment_properties()` is a very simplistic function and its result can’t be relied upon in real world use, it is present only for quick experimentation that does not need to be strictly correct.

Segment properties (script, language, and direction) should always be explicitly set by the caller. 

Script is determined using Unicode character properties (HarfBuzz has `hb_unicode_script()` to get the script of individual code points), and the caller will need to handle characters with Common and Inherited script).

Direction is determined by applying the Unicode Bidirectional Text Algorithm, any heuristic will give broken results under some circumstances.

Language can’t be reliably detected automatically, so either keep it unset or have a way for end user to set it (e.g. using some text markup).

After that, each call to HarfBuzz shape takes a text segment that has the same script, language, and direction.

**Comment by sciurius** [comment]
Thanks for the info.

I've been told that if script/language/direction has been set, `hb_buffer_guess_segment_properties` will not clobber it. If that is true then the call won't do any harm provided the user has explicitly set script/language/direction.

**Comment by khaledhosny** [comment]
> I've been told that if script/language/direction has been set, `hb_buffer_guess_segment_properties` will not clobber it.

That is true. But it gives the illusion that not setting the properties explicitly works, until it doesn’t. For example for "Hello عربي" will be guess Latin script, LTR direction and always take the language from current locale (which itself is bad, since it means changing the locale can potentially change the output).

This might be a good enough behavior for a testing tool like `hb-view` but IMHO not good behavior for a library.

**Comment by khaledhosny** [comment]
> the direction gets reset to default

I think that is the source of the misunderstanding, HarfBuzz does not have a notion of default direction (or script, or language), these are always explicitly set by the caller. The `shaper()` calls give the impression that there is a default as it calls `hb_buffer_guess_segment_properties()` which might guess right the first time.

**Comment by PhilterPaper** [comment]
> HarfBuzz does not have a notion of default direction (or script, or language), these are always explicitly set by the caller.

Well, as you point out in the next sentence, if not set explicitly, HarfBuzz will attempt to guess the properties. It's the user's fault if they mix incompatible text, such as described earlier (Hello عربي). I suppose there could be support added to help split up arbitrary text into LTR and RTL chunks (based on the alphabet in use at any given point), or even hide that from the user.

> But what if the sentences contains different typefaces (bold, italic, in sofar that exists in RTL languages)?

I would think that the _script_ used would determine RTL for bold or italic -- you still have to break up the text into smaller chunks, wherever the font changes. I don't think this should be hidden from the user at the HarfBuzz level -- perhaps Text::Layout can do the heavy lifting of splitting up long runs of text with bold and italics, hiding the small chunks that HarfBuzz needs to work with.

> I've been told that if script/language/direction has been set, hb_buffer_guess_segment_properties will not clobber it. If that is true then the call won't do any harm provided the user has explicitly set script/language/direction.

For a _single_ chunk, that is fine (the guess not overriding explicitly set properties). But what was happening before the most recent fix was that explicit settings appeared to be "sticky" and persisting into later text chunks, requiring explicit "unsetting" of that property. 

My take on it is that "guess" is fine in most cases, and explicit settings should apply only to the current text chunk, being reset for the next chunk. That way, LTR and RTL should be handled in a natural and intuitive way, with any TTB/BTT handled explicitly per text chunk. Now what happens if you have, say, Hebrew (naturally RTL) given in LTR order in the input (so far, so good), and you want to display it TTB? I assume that Arabic wouldn't be printed TTB because it's cursive and connected.

**Comment by sciurius** [comment]
Summarizing: You must always explicitly set font/size/script/language/direction before each call to shaper. Defaults as guessed by `hb_buffer_guess_segment_properties()` are not reliable even though they may seem to work in simple cases.
Harfbuzz::Shaper provides a nice service by remembering what was set and applying that to each subsequent call. That is equivalent to harfbuzz `shape()` calls with `hb_buffer_clear_contents()` in between.

So, to process a piece of text that contains of two chunk with identical settings except the font, this becomes:
````
$hb->set_font(...regular...);
$hb->set_size(...)
$hb->set_direction(...);
$hb->set_script(...)
$hb->set_language(...);
# 1st chunk
$hb->set_text("first chunk");
$hb->shaper();
# 2nd chunk, change font
$hb->set_font(...bold...);
$hb->set_text("second chunk");
$hb->shaper();
````

**Comment by PhilterPaper** [comment]
I don't think it's necessary to explicitly set script, direction, or language _unless_ HarfBuzz is guessing wrong, or you want to do something unusual (e.g., write top-to-bottom). I think most users will find that guessing of properties works just fine, so long as explicit set_* calls only apply to one text chunk (i.e., are canceled for the following chunk). Can you show a case or two where guessing produces the wrong result? I think it works well enough that explicit set_* calls should be rarely needed, which simplifies the coding for users.

If you _require_ users to explicitly set script, direction, and language (and they are "sticky" until explicitly reset), you will have to beef up the documentation. Right now it's not clear exactly which script name and which language should be used for a given font (recall the discussion about 'c' vs. 'english-united'), though direction should be pretty obvious (I need to test entering Hebrew LTR and displaying TTB). Assuming that the font files themselves are following the correct standards, we know that various OS's and possibly HarfBuzz itself are **not**, at least for the language property.

I'm going to wait until the dust has settled and you officially release 0.023 before updating PDF::Builder's textHS() Shaper-using system.

**Comment by khaledhosny** [comment]
You have been given an expert advice but you continue to ignore it and instead rely on your limited experience with multilingual text layout.

**Comment by sciurius** [comment]
I have no intention to remove the `hb_buffer_guess_segment_properties()`. As you point out it seems to do a decent job in many situations.

Automatically resetting doesn't seem right. You want to be able to do something similar to:
````
$hb->set_text("ضخ");
$hb->shaper();
print $hb->get_language, "\n";
print $hb->get_direction, "\n";
````
and have it print `Arab` and `rtl`.

If you do not want the settings to stick, just add `$hb->reset` after the call to `shaper`.

**Comment by PhilterPaper** [comment]
> As you point out it seems to do a decent job in some situations. Just don't trust on it.

Exactly! It works well enough that your default can be to let it do its thing, and only override it with explicit set_* calls when necessary. So long as there's a mechanism for canceling explicit sets when needed (such as `reset`), that works for me. I'd be a bit happier with automatic cancellation after one chunk, but I can live with a `reset` call -- it's not onerous.

>  You have been given an expert advice but you continue to ignore it and instead rely on your limited experience with multilingual text layout.

_Your_ expert advice? I keep asking for examples where the "guess" routine leads us astray, but have received none. Anyway, you are welcome to document incorrect actions by HarfBuzz::Shaper and PDF::Builder after the next release of both (incorporating the latest HarfBuzz::Shaper changes).

**Comment by sciurius** [comment]
Phil, please take your rants elsewhere. This has nothing to do with Harfbuzz::Shaper.

**Comment by PhilterPaper** [comment]
Rant? I'm not ranting. The other guy is.
